File: problems/CSES/Sorting_and_Searching/collecting_numbers_ii/collecting_numbers_ii.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, m = map(int, input().split())
    list_x = list(map(int, input().split()))

    pos = [0] * (n + 1)

    for i, x in enumerate(list_x):
        pos[x] = i

    res = 1
    for i in range(2, n + 1):
        if pos[i] < pos[i - 1]:
            res += 1

    for _ in range(m):
        l, r = map(int, input().split())
        l -= 1
        r -= 1

        x = list_x[l]
        y = list_x[r]

        pairs = set()

        for v in (x, y):
            if v > 1:
                pairs.add((v - 1, v))
            if v < n:
                pairs.add((v, v + 1))

        for u, v in pairs:
            if pos[v] < pos[u]:
                res -= 1

        list_x[l], list_x[r] = list_x[r], list_x[l]
        pos[x], pos[y] = pos[y], pos[x]

        for u, v in pairs:
            if pos[v] < pos[u]:
                res += 1

        print(res)

main()

File: problems/CSES/Sorting_and_Searching/collecting_numbers_ii/collecting_numbers_ii.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<int> list_x(n);
    vector<int> pos(n + 1);

    for (int i = 0; i < n; i++) {
        cin >> list_x[i];
        pos[list_x[i]] = i;
    }

    int res = 1;
    for (int i = 2; i < n + 1; i++) {
        if (pos[i] < pos[i - 1]) res++;
    }

    for (int i = 0; i < m; i++) {
        int left, right;
        cin >> left >> right;
        left -= 1;
        right -= 1;

        int x = list_x[left];
        int y = list_x[right];

        set<pair<int, int>> pairs;
        for (int v : {x, y}) {
            if (v > 1) pairs.insert({v - 1, v});
            if (v < n) pairs.insert({v, v + 1});
        }

        for (auto p : pairs) {
            if (pos[p.second] < pos[p.first]) res--;
        }

        swap(list_x[left], list_x[right]);
        pos[x] = right;
        pos[y] = left;

        for (auto p : pairs) {
            if (pos[p.second] < pos[p.first]) res++;
        }

        cout << res << "\n";
    }

    return 0;
}
